kaldb/src/test/java/com/slack/kaldb/clusterManager/ReplicaCreatorServiceTest.java

```diff
@@ -0,0 +1,164 @@
+    replicaCreatorService.startAsync();
+    replicaCreatorService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    assertThat(MetricsUtil.getCount(ReplicaCreatorService.REPLICAS_CREATED, meterRegistry))
+        .isEqualTo(0);
+
+    snapshotMetadataStore.close();
+    replicaMetadataStore.close();
+
+    metadataStoreService.stopAsync();
+    metadataStoreService.awaitTerminated(DEFAULT_START_STOP_DURATION);
+  }
+
+  @Test
+  public void shouldCreateReplicasIfNoneExist() throws Exception {
+    KaldbConfigs.ZookeeperConfig zkConfig =
+        KaldbConfigs.ZookeeperConfig.newBuilder()
+            .setZkConnectString(testingServer.getConnectString())
+            .setZkPathPrefix("shouldCreateReplicasIfNoneExist")
+            .setZkSessionTimeoutMs(1000)
+            .setZkConnectionTimeoutMs(1000)
+            .setSleepBetweenRetriesMs(1000)
+            .build();
+
+    MetadataStoreService metadataStoreService = new MetadataStoreService(meterRegistry, zkConfig);
+    metadataStoreService.startAsync();
+    metadataStoreService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    SnapshotMetadataStore snapshotMetadataStore =
+        new SnapshotMetadataStore(metadataStoreService.getMetadataStore(), true);
+
+    SnapshotMetadata snapshotA =
+        new SnapshotMetadata(
+            "a", "a", Instant.now().toEpochMilli() - 1, Instant.now().toEpochMilli(), 0, "a");
+    SnapshotMetadata snapshotB =
+        new SnapshotMetadata(
+            "b", "b", Instant.now().toEpochMilli() - 1, Instant.now().toEpochMilli(), 0, "b");
+
+    snapshotMetadataStore.createSync(snapshotA);
+    snapshotMetadataStore.createSync(snapshotB);
+
+    ReplicaCreatorService replicaCreatorService =
+        new ReplicaCreatorService(metadataStoreService, REPLICAS_PER_SNAPSHOT_TEST, meterRegistry);
+    replicaCreatorService.startAsync();
+    replicaCreatorService.awaitRunning(DEFAULT_START_STOP_DURATION);
+
+    await()
+        .until(
+            () ->
+                MetricsUtil.getCount(ReplicaCreatorService.REPLICAS_CREATED, meterRegistry)
+                    == REPLICAS_PER_SNAPSHOT_TEST * 2);
+
+    ReplicaMetadataStore replicaMetadataStore =
+        new ReplicaMetadataStore(metadataStoreService.getMetadataStore(), true);
+    assertThat(replicaMetadataStore.getCached().size()).isEqualTo(4);
+    assertThat(
+            (int)
+                replicaMetadataStore
+                    .getCached()
+                    .stream()
+                    .filter(replicaMetadata -> Objects.equals(replicaMetadata.snapshotId, "a"))
+                    .count())
+        .isEqualTo(REPLICAS_PER_SNAPSHOT_TEST);
+    assertThat(
+            (int)
+                replicaMetadataStore
+                    .getCached()
+                    .stream()
+                    .filter(replicaMetadata -> Objects.equals(replicaMetadata.snapshotId, "b"))
+                    .count())
+        .isEqualTo(REPLICAS_PER_SNAPSHOT_TEST);
+
+    replicaMetadataStore.close();
```

Comment:
Can we move the close operations into tearDown?

Reply:
Done.